### PR TITLE
systemd-analyze: add a new option to generate JSON output for security analysis table

### PR DESCRIPTION
**Pull request by maanyagoenka**
The format is a JSON array with objects containing the following fields: SET which indicates if the id has been set or not,
NAME which is what is used to refer to the id, DESCRIPTION which is an outline of the id state, and EXPOSURE which is an unsigned integer in the range 0.0..10.0, where a higher value corresponds to a higher security threat. The JSON version of the table is printed on the standard output file. 

The commits from the PR #[20525](https://github.com/systemd/systemd/pull/20525) have been included in this PR too because they are yet to be merged.

### CODE REVIEW

File: src/analyze/analyze.c [hunk 1]
```diff
@@ -96,6 +96,7 @@ static bool arg_offline = false;
 static unsigned arg_threshold = 100;
 static unsigned arg_iterations = 1;
 static usec_t arg_base_time = USEC_INFINITY;
+static bool arg_json_output = false;
```
Comment (bluca) [inline src/analyze/analyze.c]: This should be of type JsonFormatFlags instead, ie:

`JsonFormatFlags arg_json_format_flags = JSON_FORMAT_OFF;`

The `parse_json_argument()` function can be used to parse it

File: src/analyze/analyze.c [hunk 2]
```diff
@@ -2234,6 +2236,8 @@ static int help(int argc, char *argv[], void *userdata) {
                "     --version               Show package version\n"
                "     --security-policy=PATH  Obtain a set of custom requirements\n"
                "                             against which to compare the unit file(s)\n"
+               "     --json-output=BOOL      Generate JSON output of the security\n"
```
Comment (bluca) [inline src/analyze/analyze.c]: should be called just `json` like other programs:

```
               "     --json=pretty|short|off\n"
               "                          Generate JSON output..."
```

File: src/analyze/analyze-security.c [hunk 3]
```diff
@@ -1800,6 +1801,16 @@ static int assess(const SecurityInfo *info,
                         return table_log_add_error(r);
         }
 
+        if (json_output && details_table) {
```
Comment (bluca) [inline src/analyze/analyze-security.c]: this check can become `!(json_flags & JSON_FORMAT_OFF)` and then `json_flags` should be passed to `json_variant_dump` instead of hard-coding JSON_FORMAT_PRETTY

Comment (bluca) [reply src/analyze/analyze-security.c]: and the block above that does the printf of the table should have the opposite check, as --json should stop the non-json output from being printed

Comment (bluca) [reply src/analyze/analyze-security.c]: (same for overview_table in analyze_security())

File: shell-completion/bash/systemd-analyze [hunk 4]
```diff
@@ -144,7 +144,7 @@ _systemd_analyze() {
 
     elif __contains_word "$verb" ${VERBS[SECURITY]}; then
         if [[ $cur = -* ]]; then
-            comps='--help --version --no-pager --system --user -H --host -M --machine --offline --threshold --security-policy'
+            comps='--help --version --no-pager --system --user -H --host -M --machine --offline --threshold --security-policy --json'
```
Comment (bluca) [inline shell-completion/bash/systemd-analyze]: --json=off --json=short --json=pretty

File: src/analyze/analyze.c [hunk 5]
```diff
@@ -2234,6 +2236,8 @@ static int help(int argc, char *argv[], void *userdata) {
                "     --version               Show package version\n"
                "     --security-policy=PATH  Obtain a set of custom requirements\n"
                "                             against which to compare the unit file(s)\n"
+               "     --json=MODE             Generate JSON output of the security\n"
```
Comment (bluca) [inline src/analyze/analyze.c]: let's be explicit here like other tools:

```
               "     --json=pretty|short|off\n"
               "                          Generate JSON output of..."
```

File: src/analyze/analyze-security.c [hunk 6]
```diff
@@ -1624,7 +1625,7 @@ static int assess(const SecurityInfo *info,
         int r;
 
         if (!FLAGS_SET(flags, ANALYZE_SECURITY_SHORT)) {
-                details_table = table_new(" ", "name", "description", "weight", "badness", "range", "exposure");
+                details_table = table_new("set", "name", "description", "weight", "badness", "range", "exposure");
```
Comment (poettering) [inline src/analyze/analyze-security.c]: please do not modify the traditional console output

File: src/analyze/analyze-security.c [hunk 7]
```diff
@@ -1800,6 +1803,16 @@ static int assess(const SecurityInfo *info,
...
         }
 
+        if (!(json_format_flags & JSON_FORMAT_OFF)) {
+                _cleanup_(json_variant_unrefp) JsonVariant *json_table = NULL;
+
+                r = table_to_json(details_table, &json_table);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to convert table to JSON: %m");
+
+                json_variant_dump(json_table, json_format_flags, stdout, /* prefix= */ NULL);
```
Comment (poettering) [inline src/analyze/analyze-security.c]: why did you move output of this table here if in JSON mode? why not output it where it is output in non-JSON mode? i.e. where the table_print() is currently placed above?

Also, please use table_print_with_pager() these days. which will handle both JSON and non-JSON output for you automatically

**Comment by maanyagoenka** [comment]
@bluca how do I move all the changes from the previous PR (the unmerged ones in PR: systemd-analyze: add option to enable users to custom define security requirements in the form of a .json file #20525) to this one while not changing anything in the previous PR?

**Comment by bluca** [comment]
> @bluca how do I move all the changes from the previous PR (the unmerged ones in PR: systemd-analyze: add option to enable users to custom define security requirements in the form of a .json file #20525) to this one while not changing anything in the previous PR?
> Or I guess I could just wait for it to get merged and then resolve when conflicts arise

From this branch, something like: git rebase --onto other-branch-name HEAD^ will move the latest commit on top of the content of the other branch

**Comment by bluca** [comment]
Needs a manual rebase

File: src/analyze/analyze-security.c [hunk 8]
```diff
@@ -1782,15 +1789,15 @@ static int assess(const SecurityInfo *info,
                                 description = access_description_na(a, policy);
                                 color = NULL;
                         } else if (badness == a->range) {
-                                checkmark = special_glyph(SPECIAL_GLYPH_CROSS_MARK);
+                                checkmark = ((json_format_flags == JSON_FORMAT_OFF) ? special_glyph(SPECIAL_GLYPH_CROSS_MARK) : "no");
```
Comment (poettering) [inline src/analyze/analyze-security.c]: as mentioned in my earlier review: please rework format-table.[ch] to get a new field table TABLE_BOOLEAN_CHECKMARK which would be very much like TABLE_BOOLEAN, but instead of displaying yes/no for booleans in console output, would chow cross mark/check mark instead — and real booleans in JSON mode. Then, rework this code here to use that instead of TABLE_STRING for the second column.

Or in othewords: please push the booleans here down into the table formatter, and add the logic to display the right glyphs for them to the table formatter code, instead of here, ahead of time.

File: src/analyze/analyze-security.c [hunk 9]
```diff
@@ -1846,7 +1853,7 @@ static int assess(const SecurityInfo *info,
                                 return log_error_errno(r, "Failed to update cell in table: %m");
                 }
 
-                r = table_print(details_table, stdout);
+                r =  table_print_with_pager(details_table, json_format_flags, /* pager_flags= */ 0, /* show_header */ true);
```
Comment (poettering) [inline src/analyze/analyze-security.c]: spurious space

File: src/analyze/analyze-security.c [hunk 10]
```diff
@@ -2805,10 +2815,9 @@ int analyze_security(sd_bus *bus,
                         fflush(stdout);
                 }
 
-                r = table_print(overview_table, stdout);
+                r = table_print_with_pager(overview_table, json_format_flags, /* pager_flags= */ 0, /* show_header= */ true);
```
Comment (poettering) [inline src/analyze/analyze-security.c]: please pass `arg_pager_flags` from the main program through here too, instead of passing them as 0 here. (and similar above)

File: src/analyze/analyze-security.c [hunk 11]
```diff
@@ -1732,7 +1733,13 @@ static int assess(const SecurityInfo *info,
...
 
         if (!FLAGS_SET(flags, ANALYZE_SECURITY_SHORT)) {
-                details_table = table_new(" ", "name", "description", "weight", "badness", "range", "exposure");
+                const char *first_column_name = " ";
+
+                /* The json output of the table should have a proper key name for the first field */
+                if (json_format_flags != JSON_FORMAT_OFF)
+                        first_column_name = "set";
+
+                details_table = table_new(first_column_name, "name", "description", "weight", "badness", "range", "exposure");
```
Comment (poettering) [inline src/analyze/analyze-security.c]: I think in JSON mode the "name" column should get the assesor's `json_field` field instead of `id`. i.e. it's important the policy JSON and the output you generate here can be reasonably mapped to each other.

**Comment by bluca** [comment]
Also needs rebase

File: src/analyze/analyze-security.c (hunk 11 above)
Comment (maanyagoenka) [reply src/analyze/analyze-security.c]: This was addressed on lines 1808 and 1809 of the analyze-security.c file

File: src/analyze/analyze-security.c (hunk 8 above)
Comment (maanyagoenka) [reply src/analyze/analyze-security.c]: @poettering @bluca I'm not sure I completely understand. How would this new field table know when json_format_flags has been set to be JSON_FORMAT_OFF and when it is set with the other two values (pretty | short). Or are you looking for something like this:

    r = table_add_many(details_table,
                                           (json_format_flags == JSON_FORMAT_OFF) ? TABLE_BOOLEAN_CHECKMARK: 
                                            TABLE_BOOLEAN, checkmark,
                                           TABLE_SET_MINIMUM_WIDTH, 1,
                                            .....);

And what is checkmark's value set to in the if else statements above. The original before my change was: 
    
    else if (badness == a->range) {
                                checkmark = special_glyph(SPECIAL_GLYPH_CROSS_MARK);
                                ......
    }

Comment (poettering) [reply src/analyze/analyze-security.c]: if you look in format-table.c you'll find two very distinct code paths: table_data_to_json() generates json objects from the table content. here the new TABLE_BOOLEAN_CHECKMARK should have the exact same codepaths as TABLE_BOOLEAN. And then there is table_data_format() generates text to output if text output is requested, here you should use different codepaths for TABLE_BOOLEAN and TABLE_BOOLEAN_CHECKMARK.

table_print_with_pager() already contains the right logic that dispatches things towards one of the codepaths mentioned above in JSON mode, and to the other in text mode, you don't have to take care of that anymore, it already gets a json_format_flags parameter to figure that out.

**Comment by maanyagoenka** [comment]
> Also needs rebase

Can't figure out why these tests are failings. Says there's an optimization error?

File: src/analyze/analyze.c [hunk 12]
```diff
@@ -2169,7 +2170,7 @@ static int do_security(int argc, char *argv[], void *userdata) {
         (void) pager_open(arg_pager_flags);
 
         if (arg_security_policy) {
-                r = json_parse_file(/*f=*/ NULL, arg_security_policy, /*flags=*/ 0, &policy, &line, &column);
+                r = json_parse_file(/*f=*/ NULL, arg_security_policy, arg_pager_flags, &policy, &line, &column);
```
Comment (poettering) [inline src/analyze/analyze.c]: JsonParseFlags is not the same as PagerFlags

File: src/analyze/analyze.c [hunk 13]
```diff
@@ -2181,7 +2182,7 @@ static int do_security(int argc, char *argv[], void *userdata) {
                         return r;
 
                 if (f != NULL) {
-                        r = json_parse_file(f, pp, /*flags=*/ 0, &policy, &line, &column);
+                        r = json_parse_file(f, pp, arg_pager_flags, &policy, &line, &column);
```
Comment (poettering) [inline src/analyze/analyze.c]: ditto

File: src/shared/format-table.c [hunk 14]
```diff
@@ -262,6 +262,9 @@ static size_t table_data_size(TableDataType type, const void *data) {
         case TABLE_BOOLEAN:
                 return sizeof(bool);
 
+        case TABLE_BOOLEAN_CHECKMARK:
```
Comment (poettering) [inline src/shared/format-table.c:270]: merge with the preceding block, i.e. see how TABLE_TIMESTAMP/TABLE_TIMESTAMP_UTC/… do it

File: src/shared/format-table.c [hunk 15]
```diff
@@ -845,6 +848,11 @@ int table_add_many_internal(Table *t, TableDataType first_type, ...) {
                         data = &buffer.b;
                         break;
 
+                case TABLE_BOOLEAN_CHECKMARK:
+                        buffer.b = ((va_arg(ap, const char *) == special_glyph(SPECIAL_GLYPH_CROSS_MARK)) ? false : true);
+                        data = &buffer.b;
+                        break;
```
Comment (poettering) [inline src/shared/format-table.c]: nope, this should do the same thing as TABLE_BOOLEAN, i.e. expect an `int`. i.e. just add a `case TABLE_BOOLEAN_CHECKMARK:` as one extra line after the `case TABLE_BOOLEAN:`, that's all.

i.e. TABLE_BOOLEAN and TABLE_BOOLEAN_CHECKMARK should be read, processed, stored exacly the same as TABLE_BOOLEAN in almost all ways — except one: when *outputting* in *text* form, show the cross mark, and not the yes/no thing.

File: src/analyze/analyze-security.c [hunk 16]
```diff
@@ -2589,7 +2599,7 @@ static int get_security_info(Unit *u, ExecContext *c, CGroupContext *g, Security
         return 0;
 }
 
-static int offline_security_check(Unit *u, unsigned threshold, JsonVariant *policy) {
+static int offline_security_check(Unit *u, unsigned threshold, JsonVariant *policy, PagerFlags pager_flags, JsonFormatFlags json_format_flags) {
```
Comment (poettering) [inline src/analyze/analyze-security.c]: this line is so long now, please line break

File: src/analyze/analyze-security.c [hunk 17]
```diff
@@ -1846,7 +1854,7 @@ static int assess(const SecurityInfo *info,
                                 return log_error_errno(r, "Failed to update cell in table: %m");
                 }
 
-                r = table_print(details_table, stdout);
+                r =  table_print_with_pager(details_table, json_format_flags, pager_flags, /*show_header=*/ true);
```
Comment (poettering) [inline src/analyze/analyze-security.c]: please remove spurious duplicate space after the `=`

File: src/analyze/analyze-security.c [hunk 18]
```diff
@@ -1798,11 +1806,11 @@ static int assess(const SecurityInfo *info,
                         if (d)
                                 description = d;
 
-                        if (json_variant_by_key(policy, a->json_field) != NULL)
+                        if (json_variant_by_key(policy, a->json_field) != NULL || json_format_flags != JSON_FORMAT_OFF)
                                 id = a->json_field;
```
Comment (poettering) [inline src/analyze/analyze-security.c]: actually i think we should always add both `a->json_field` *and* `a->id` to the table, in separate columns, and include both in the JSON output, but hide the json field name from the regular text output, using `table_set_display()`.

i.e. it's a good thing if the table is always filled the same way, and we simply tweak what parts we show of it after it is put together depending on JSON or not.

File: man/systemd-analyze.xml [hunk 19]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
+        <listitem><para>With <command>security</command> command, generate a JSON formatted
+        output of the security analysis table. The format is a JSON array with objects
+        containing the following fields: SET which indicates if the id has been set or not,
+        NAME which is what is used to refer to the id, DESCRIPTION which is an outline of the
+        id state, and EXPOSURE which is an unsigned integer in the range 0.0..10.0, where a
+        higher value corresponds to a higher security threat. The JSON version of the table is
+        printed on the standard output file. The <replaceable>MODE</replaceable> passed to the
+        option can be one of three: off which is the default, pretty which uses flags
+        JSON_FORMAT_PRETTY and JSON_FORMAT_COLOR_AUTO, and short which uses flag
+        JSON_FORMAT_NEWLINE.</para></listitem>
```
Comment (poettering) [inline man/systemd-analyze.xml]: this isn't right. the uppercase flags names are not exposed on the cmdline

File: man/systemd-analyze.xml [hunk 20]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
 
 
+      <varlistentry>
+        <term><option>--json=<replaceable>MODE</replaceable></option></term>
+
+        <listitem><para>With <command>security</command> command, generate a JSON formatted
+        output of the security analysis table. The format is a JSON array with objects
+        containing the following fields: SET which indicates if the id has been set or not,
+        NAME which is what is used to refer to the id, DESCRIPTION which is an outline of the
+        id state, and EXPOSURE which is an unsigned integer in the range 0.0..10.0, where a
```
Comment (poettering) [inline man/systemd-analyze.xml]: the json fields are named quite differently now.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/analyze/analyze.c (hunk 12 above)
Comment (bluca) [reply src/analyze/analyze.c]: @maanyagoenka this is what's causing the build failure

File: src/analyze/analyze-security.c [hunk 21]
```diff
@@ -1732,7 +1734,13 @@ static int assess(const SecurityInfo *info,
         int r;
 
         if (!FLAGS_SET(flags, ANALYZE_SECURITY_SHORT)) {
-                details_table = table_new(" ", "name", "description", "weight", "badness", "range", "exposure");
+                const char *first_column_name = " ";
+
+                /* The json output of the table should have a proper key name for the first field */
+                if (json_format_flags != JSON_FORMAT_OFF)
+                        first_column_name = "set";
```
Comment (poettering) [inline src/analyze/analyze-security.c]: to handle this a bit nicer i have prepped #20630 now. Plese rebase your work on top of it. 

With that you can leave the table_new() entirely unmodified, but then add a line:

```c
r = table_set_json_field_name(t, 0, "set");
if (r < 0)
        return log_error_errno(r, "Failed to set JSON field name of column 0: %m");
```

That makes things a bit nicer conceptually, since that way you'll store always the same data in the table. And only as very last step, when outputting things you go two different paths for json vs. regular mode.

**Comment by poettering** [comment]
@bluca would be great if you could have a look at #20630, so that we can merge it quickly and @maanyagoenka can rebase their work on it. see above for an explanation why I think that would be good.

**Comment by bluca** [comment]
> @bluca would be great if you could have a look at #20630, so that we can merge it quickly and @maanyagoenka can rebase their work on it. see above for an explanation why I think that would be good.

thank you, looks great!

File: src/analyze/analyze-security.c (hunk 18 above)
Comment (maanyagoenka) [reply src/analyze/analyze-security.c:1810]: How does table_set_display() work because it seems like adding these lines to the code is just producing an 8 column table with columns: " ", id, json_field, description, exposure, " ", id, description, exposure.

    if (json_format_flags & JSON_FORMAT_OFF)
                        r = table_set_display(details_table, (size_t) 0, (size_t) 1, (size_t) 3, (size_t) 7);
                else
                        r = table_set_display(details_table, (size_t) 0, (size_t) 1, (size_t) 2, (size_t) 3, (size_t) 7);
                if (r < 0)
                        return log_error_errno(r, "Failed to set columns to display: %m");

And the json_field seems to be included for both the off and non off cases of JSON. What am I missing?

File: man/systemd-analyze.xml [hunk 22]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
       </varlistentry>
 
 
+      <varlistentry>
+        <term><option>--json=<replaceable>MODE</replaceable></option></term>
+
+        <listitem><para>With <command>security</command> command, generate a JSON formatted
+        output of the security analysis table. The format is a JSON array with objects
+        containing the following fields: set which indicates if the id has been set or not,
```
Comment (poettering) [inline man/systemd-analyze.xml]: enclose  the field names in some suitable docbook tag. maybe `<varname>`.

Comment (poettering) [reply man/systemd-analyze.xml]: if the id has been set or not → "if the setting has been enabled or not"

File: man/systemd-analyze.xml [hunk 23]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
+        name which is what is used to refer to the id, json_field which is the JSON compatible
```
Comment (poettering) [inline man/systemd-analyze.xml]: the id → the setting

File: man/systemd-analyze.xml [hunk 24]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
+        name of the id, description which is an outline of the id state, and exposure which
```
Comment (poettering) [inline man/systemd-analyze.xml]: name of the id → identifier of the setting

Comment (poettering) [reply man/systemd-analyze.xml]: id state → setting state

File: man/systemd-analyze.xml [hunk 25]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
+        is an unsigned integer in the range 0.0..10.0, where a higher value corresponds to a
```
Comment (poettering) [inline man/systemd-analyze.xml]: you specify the range with one digit after the dot accuracy, but talk about integers. That's weird.

File: man/systemd-analyze.xml [hunk 26]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
+        higher security threat. The JSON version of the table is printed on the standard output
+        file. The <replaceable>MODE</replaceable> passed to the option can be one of three:
```
Comment (poettering) [inline man/systemd-analyze.xml]: standard output is not really a file (i mean, it can be, but usually is not). Just call it "standard output", make no mention of "file".

File: man/systemd-analyze.xml [hunk 27]
```diff
@@ -1109,6 +1109,21 @@ Service b@0.service not loaded, b.socket cannot be started.
...
+        off which is the default, and pretty and short which use JSON format flags to
```
Comment (poettering) [inline man/systemd-analyze.xml]: enclose off/pretty/short in `<option>` or so

File: src/analyze/analyze-security.c [hunk 28]
```diff
@@ -1775,7 +1781,6 @@ static int assess(const SecurityInfo *info,
 
                 if (details_table) {
                         const char *checkmark, *description, *color = NULL;
```
Comment (poettering) [inline src/analyze/analyze-security.c]: please change `checkmark` to be of type `int`here. Then assign -1 if badness is UINT64_MAX, 0 if cross mark, and 1 if check mark. 

Then add a TABLE_EMPTY cell from it if negative, otherwise TABLE_BOOLEAN_CHECKMARK

File: src/analyze/analyze-security.c [hunk 29]
```diff
@@ -1798,16 +1803,14 @@ static int assess(const SecurityInfo *info,
...
                         r = table_add_many(details_table,
-                                           TABLE_STRING, checkmark,
+                                           TABLE_BOOLEAN_CHECKMARK, checkmark == special_glyph(SPECIAL_GLYPH_CHECK_MARK),
                                            TABLE_SET_MINIMUM_WIDTH, 1,
                                            TABLE_SET_MAXIMUM_WIDTH, 1,
                                            TABLE_SET_ELLIPSIZE_PERCENT, 0,
                                            TABLE_SET_COLOR, color,
-                                           TABLE_STRING, id, TABLE_SET_URL, a->url,
+                                           TABLE_STRING, a->id, TABLE_STRING, a->json_field,
+                                           TABLE_SET_URL, a->url,
```
Comment (poettering) [inline src/analyze/analyze-security.c]: the URL should be assigned to the `id` colun, not the `json_field` column. hence insert the new line *after* the TABLE_SET_URL thing, not before

File: src/analyze/analyze.c [hunk 30]
```diff
@@ -2169,7 +2170,7 @@ static int do_security(int argc, char *argv[], void *userdata) {
         (void) pager_open(arg_pager_flags);
 
         if (arg_security_policy) {
-                r = json_parse_file(/*f=*/ NULL, arg_security_policy, /*flags=*/ 0, &policy, &line, &column);
+                r = json_parse_file(/*f=*/ NULL, arg_security_policy, /*flags=*/0, &policy, &line, &column);
```
Comment (poettering) [inline src/analyze/analyze.c]: why this change?

File: src/shared/format-table.c [hunk 31]
```diff
@@ -1443,6 +1445,9 @@ static const char *table_data_format(Table *t, TableData *d, bool avoid_uppercas
         case TABLE_BOOLEAN:
                 return yes_no(d->boolean);
 
+        case TABLE_BOOLEAN_CHECKMARK:
+                return d->boolean ? special_glyph(SPECIAL_GLYPH_CHECK_MARK) : special_glyph(SPECIAL_GLYPH_CROSS_MARK);
```
Comment (poettering) [inline src/shared/format-table.c]: maybe write as:

```c
return special_glyph(d->boolean ? SPECIAL_GLYPH_CHECK_MARK : SPECIAL_GLYPH_CROSS_MARK);
```

which is identical, but makes the common code between the two codepaths more comprehensive

File: src/analyze/analyze.c [hunk 32]
```diff
@@ -2181,7 +2182,7 @@ static int do_security(int argc, char *argv[], void *userdata) {
                         return r;
 
                 if (f != NULL) {
-                        r = json_parse_file(f, pp, /*flags=*/ 0, &policy, &line, &column);
+                        r = json_parse_file(f, pp, /*flags=*/0, &policy, &line, &column);
```
Comment (poettering) [inline src/analyze/analyze.c]: ditto

**Review by poettering — CHANGES_REQUESTED** [review]
close

**Comment by bluca** [comment]
> close

Updated, ptal! (note that Maanya's internship finished last week, so I'll see this one through)
